src/core_atmosphere/physics/physics_monan/module_bl_mixingscalars.F

```diff
@@ -0,0 +1,305 @@
+          call mix_scalars_opt(kts,kte,dt_pbl,dz,rho,dfh,nscalars,local_scalars)
+
+          !--- copy back the updated scalars 
+          !--- (a tendency could be computed here by saving the previous values of the scalars)
+          do isc = 1,nscalars
+              scalars(index_scalar(isc),kts:kte,i) = local_scalars(kts:kte,isc)  
+          enddo 
+
+      enddo ! loop i
+
+      if(allocated(local_scalars)) deallocate(local_scalars)
+
+   end subroutine driver_pbl_scalars
+!-----------------------------------------------------------------------------
+
+   subroutine mix_scalars(kts,kte,delt,dz,rho,dfh,nscalars,local_scalars)
+     integer, intent(in) :: kts,kte,nscalars
+     real(kind_phys), intent(in) :: delt
+     real(kind_phys), dimension(kts:kte), intent(in) :: dfh,dz,rho
+     real(kind_phys), dimension(kts:kte,nscalars), intent(inout) :: local_scalars
+
+!--- local vars
+     integer                               :: k,isc
+     real(kind_phys)                       :: dztop
+     real(kind_phys), dimension(kts:kte)   :: dtz
+     real(kind_phys), dimension(kts:kte)   :: a,b,c,d,x
+     real(kind_phys), dimension(kts:kte)   :: rhoinv
+     real(kind_phys), dimension(kts:kte+1) :: rhoz,khdz
+
+     dztop = 0.5*(dz(kte)+dz(kte-1))
+
+     do k = kts,kte
+        dtz(k)=delt/dz(k)
+     enddo
+     
+     !---  Prepare "constants" for diffusion equation.
+     !---  khdz = rho*Kh/dz = rho*dfh
+     rhoz(kts)  =rho(kts)
+     rhoinv(kts)=1./rho(kts)
+     khdz(kts)  =rhoz(kts)*dfh(kts)
+
+     do k = kts+1,kte
+        rhoz(k)  =(rho(k)*dz(k-1) + rho(k-1)*dz(k))/(dz(k-1)+dz(k))
+        rhoz(k)  =   max(rhoz(k),1e-4)
+        rhoinv(k)=1./max( rho(k),1e-4)
+        !dzk      = 0.5*(dz(k)+dz(k-1))
+        khdz(k)  = rhoz(k)*dfh(k)
+     enddo
+     rhoz(kte+1) = rhoz(kte)
+     khdz(kte+1) = rhoz(kte+1)*dfh(kte)
+
+     do isc = 1,nscalars
+        k=kts
+
+        a(k)=  -dtz(k)*khdz(k)*rhoinv(k)
+        b(k)=1.+dtz(k)*(khdz(k+1)+khdz(k))*rhoinv(k)
+        c(k)=  -dtz(k)*khdz(k+1)*rhoinv(k)           
+        d(k)=  local_scalars(k,isc) 
+
+        do k=kts+1,kte-1
+           a(k) =   -dtz(k)*khdz(k)*rhoinv(k)     
+           b(k) = 1.+dtz(k)*(khdz(k)+khdz(k+1))*rhoinv(k)
+           c(k) =   -dtz(k)*khdz(k+1)*rhoinv(k) 
+           d(k) = local_scalars(k,isc) 
+        enddo
+        ! prescribed value at top
+        a(kte) = 0.
+        b(kte) = 1.
+        c(kte) = 0.
+        d(kte) = local_scalars(kte,isc)
+        
+        call tridiag3(kte,a,b,c,d,x)
+        
+        do k=kts,kte
+           local_scalars(k,isc) = max(1.e-12,x(k))
+        enddo
+     enddo
+
+   end subroutine mix_scalars
+!-----------------------------------------------------------------------------
+   subroutine mix_scalars_opt(kts,kte,delt,dz,rho,dfh,nscalars,local_scalars)
+     integer, intent(in) :: kts,kte,nscalars
+     real(kind_phys), intent(in) :: delt
+     real(kind_phys), dimension(kts:kte), intent(in) :: dfh,dz,rho
```

Comment:
"DTN 01 - 4.37 - The preferred mechanism for dynamic memory allocation is automatic arrays.

Example:
real(kind_phys), dimension(kts:kte), intent(in) :: dfh,dz,rho

Suggestion:
real(kind_phys), dimension(:), intent(in) :: dfh,dz,rho

Reply:
not solved.